File: Source/Other/DatabaseAbstraction/SPMySQLConnectionWrapper.m
//
//  SPMySQLConnectionWrapper.m
//  Sequel Ace
//

#import "SPMySQLConnectionWrapper.h"
#import <SPMySQL/SPMySQL.h>

@implementation SPMySQLConnectionWrapper {
    SPMySQLConnection *_connection;
}

- (instancetype)initWithConnection:(SPMySQLConnection *)connection {
    if (self = [super init]) {
        _connection = connection;
    }
    return self;
}

- (SPMySQLConnection *)underlyingConnection {
    return _connection;
}

+ (NSUInteger)defaultPort { return 3306; }

// ─── Connectivity ────────────────────────────────────────────────────────────

- (BOOL)connect    { return [_connection connect]; }
- (void)disconnect { [_connection disconnect]; }
- (BOOL)isConnected { return [_connection isConnected]; }
- (BOOL)isConnectedViaSSL { return [_connection isConnectedViaSSL]; }

// ─── Configuration ───────────────────────────────────────────────────────────

- (void)setUsername:(NSString *)u { [_connection setUsername:u]; }
- (void)setPassword:(NSString *)p { [_connection setPassword:p]; }
- (void)setHost:(NSString *)h     { [_connection setHost:h]; }
- (void)setPort:(NSUInteger)port  { [_connection setPort:port]; }
- (void)setDatabase:(NSString *)db { [_connection setDatabase:db]; }
- (void)setTimeout:(NSUInteger)t  { [_connection setTimeout:t]; }
- (void)setUseSSL:(BOOL)ssl       { [_connection setUseSSL:ssl]; }

- (void)setDelegate:(id<SPDatabaseConnectionProxy>)delegate {
    // SPMySQLConnection uses its own SPMySQLConnectionDelegate protocol.
    // SPDatabaseDocument already conforms to both, so a direct cast is safe.
    [_connection setDelegate:(id<SPMySQLConnectionDelegate>)delegate];
}

// ─── Database selection ───────────────────────────────────────────────────────

- (BOOL)selectDatabase:(NSString *)aDatabase { return [_connection selectDatabase:aDatabase]; }
- (NSArray *)databases { return [_connection databases]; }

// ─── Querying ────────────────────────────────────────────────────────────────

- (id<SPDatabaseResult>)queryString:(NSString *)query {
    return (id<SPDatabaseResult>)[_connection queryString:query];
}

- (id)streamingQueryString:(NSString *)query {
    return [_connection streamingQueryString:query];
}

- (id)streamingQueryString:(NSString *)query useLowMemoryBlockingStreaming:(BOOL)full {
    return [_connection streamingQueryString:query useLowMemoryBlockingStreaming:full];
}

- (BOOL)queryErrored { return [_connection queryErrored]; }
- (NSString *)lastErrorMessage { return [_connection lastErrorMessage]; }
- (NSUInteger)lastErrorID { return [_connection lastErrorID]; }
- (unsigned long long)rowsAffectedByLastQuery { return [_connection rowsAffectedByLastQuery]; }
- (unsigned long long)lastInsertID { return [_connection lastInsertID]; }
- (void)cancelCurrentQuery { [_connection cancelCurrentQuery]; }

// ─── Encoding ────────────────────────────────────────────────────────────────

- (NSString *)encoding { return [_connection encoding]; }
- (BOOL)setEncoding:(NSString *)enc { return [_connection setEncoding:enc]; }
- (void)storeEncodingForRestoration { [_connection storeEncodingForRestoration]; }
- (void)restoreStoredEncoding { [_connection restoreStoredEncoding]; }

- (NSString *)preferredUTF8Encoding {
    // Prefer utf8mb4 on MySQL 5.5.3+, fall back to utf8.
    if ([_connection serverVersionIsGreaterThanOrEqualTo:5 minorVersion:5 releaseVersion:3]) {
        return @"utf8mb4";
    }
    return @"utf8";
}

// ─── Server info ─────────────────────────────────────────────────────────────

- (NSString *)serverVersionString { return [_connection serverVersionString]; }
- (NSUInteger)serverMajorVersion  { return [_connection serverMajorVersion]; }
- (NSUInteger)serverMinorVersion  { return [_connection serverMinorVersion]; }
- (NSUInteger)serverReleaseVersion { return [_connection serverReleaseVersion]; }

- (BOOL)serverVersionIsGreaterThanOrEqualTo:(NSUInteger)major minorVersion:(NSUInteger)minor releaseVersion:(NSUInteger)release {
    return [_connection serverVersionIsGreaterThanOrEqualTo:major minorVersion:minor releaseVersion:release];
}

- (NSString *)getServerVariableValue:(NSString *)variable {
    NSString *query = [NSString stringWithFormat:@"SHOW VARIABLES LIKE '%@'", variable];
    id result = [_connection queryString:query];
    if (result && [result respondsToSelector:@selector(getRowAsArray)]) {
        NSArray *row = [result getRowAsArray];
        if (row.count >= 2) return row[1];
    }
    return nil;
}

// ─── Collation helpers ────────────────────────────────────────────────────────

- (NSArray *)getCollationsForEncoding:(NSString *)encoding {
    // Return empty; callers fall back to the existing MySQL-specific query path.
    return @[];
}

// ─── Backend type ─────────────────────────────────────────────────────────────

- (BOOL)isPostgreSQL { return NO; }
- (NSString *)identifierQuoteCharacter { return @"`"; }

@end

File: Source/Other/DatabaseAbstraction/SPMySQLConnectionWrapper.swift
//
//  SPMySQLConnectionWrapper.swift
//  Sequel Ace
//

import Foundation

@objc(SPMySQLConnectionWrapper)
final class SPMySQLConnectionWrapper: NSObject, SPDatabaseConnection {
    private let connection: SPMySQLConnection

    @objc(initWithConnection:)
    init(connection: SPMySQLConnection) {
        self.connection = connection
        super.init()
    }

    @objc var underlyingConnection: SPMySQLConnection { connection }

    @objc var host: String? {
        get { connection.host }
        set { connection.host = newValue }
    }

    @objc var database: String? {
        get { connection.database }
        set { connection.database = newValue }
    }

    @objc static var defaultPort: UInt { 3306 }

    func connect() -> Bool { connection.connect() }
    func disconnect() { connection.disconnect() }
    func isConnected() -> Bool { connection.isConnected() }
    func isConnectedViaSSL() -> Bool { connection.isConnectedViaSSL() }

    func setUsername(_ username: String) { connection.username = username }
    func setPassword(_ password: String) { connection.password = password }
    func setPort(_ port: UInt) { connection.port = port }
    func setTimeout(_ timeout: UInt) { connection.timeout = timeout }
    func setUseSSL(_ useSSL: Bool) { connection.useSSL = useSSL }

    func setDelegate(_ delegate: SPDatabaseConnectionProxy?) {
        connection.setDelegate(delegate as? SPMySQLConnectionDelegate)
    }

    func selectDatabase(_ aDatabase: String) -> Bool { connection.selectDatabase(aDatabase) }
    func databases() -> [String] { connection.databases() as? [String] ?? [] }

    func queryString(_ query: String) -> SPDatabaseResult? {
        connection.queryString(query) as? SPDatabaseResult
    }

    func streamingQueryString(_ query: String) -> Any? {
        connection.streamingQueryString(query)
    }

    func streamingQueryString(_ query: String, useLowMemoryBlockingStreaming fullStreaming: Bool) -> Any? {
        connection.streamingQueryString(query, useLowMemoryBlockingStreaming: fullStreaming)
    }

    func queryErrored() -> Bool { connection.queryErrored() }
    func lastErrorMessage() -> String? { connection.lastErrorMessage() }
    func lastErrorID() -> UInt { connection.lastErrorID() }
    func rowsAffectedByLastQuery() -> UInt64 { connection.rowsAffectedByLastQuery() }
    func lastInsertID() -> UInt64 { connection.lastInsertID() }
    func cancelCurrentQuery() { connection.cancelCurrentQuery() }

    func encoding() -> String? { connection.encoding() }
    func setEncoding(_ encoding: String) -> Bool { connection.setEncoding(encoding) }
    func storeEncodingForRestoration() { connection.storeEncodingForRestoration() }
    func restoreStoredEncoding() { connection.restoreStoredEncoding() }

    func preferredUTF8Encoding() -> String {
        if connection.serverVersionIsGreaterThanOrEqual(to: 5, minorVersion: 5, releaseVersion: 3) {
            return "utf8mb4"
        }
        return "utf8"
    }

    func serverVersionString() -> String? { connection.serverVersionString() }
    func serverMajorVersion() -> UInt { connection.serverMajorVersion() }
    func serverMinorVersion() -> UInt { connection.serverMinorVersion() }
    func serverReleaseVersion() -> UInt { connection.serverReleaseVersion() }

    func serverVersionIsGreaterThanOrEqual(to major: UInt, minorVersion minor: UInt, releaseVersion release: UInt) -> Bool {
        connection.serverVersionIsGreaterThanOrEqual(to: major, minorVersion: minor, releaseVersion: release)
    }

    func getServerVariableValue(_ variable: String) -> String? {
        let query = "SHOW VARIABLES LIKE '\(variable)'"
        guard let result = connection.queryString(query),
              let row = result.getRowAsArray() as? [Any],
              row.count >= 2 else { return nil }
        return row[1] as? String
    }

    func getCollationsForEncoding(_ encoding: String) -> [Any] { [] }

    func isPostgreSQL() -> Bool { false }
    func identifierQuoteCharacter() -> String { "`" }

    @objc func isMariaDB() -> Bool { connection.isMariaDB() }

    @objc func escapeAndQuoteString(_ theString: String) -> String? {
        connection.escapeAndQuoteString(theString)
    }

    @objc var lastQueryWasCancelled: Bool {
        connection.lastQueryWasCancelled
    }
}
